fendermint/testing/materializer/src/docker/node.rs

```diff
@@ -0,0 +1,566 @@
+pub struct DockerNode {
+    /// Logical name of the node in the subnet hierarchy.
+    node_name: NodeName,
+    fendermint: DockerContainer,
+    cometbft: DockerContainer,
+    ethapi: Option<DockerContainer>,
+    port_range: DockerPortRange,
+    path: PathBuf,
+}
+
+impl DockerNode {
+    pub async fn get_or_create<'a>(
+        root: impl AsRef<Path>,
+        dh: DockerWithDropHandle,
+        node_name: &NodeName,
+        node_config: NodeConfig<'a, DockerMaterials>,
+        port_range: DockerPortRange,
+    ) -> anyhow::Result<Self> {
+        let fendermint_name = container_name(node_name, "fendermint");
+        let cometbft_name = container_name(node_name, "cometbft");
+        let ethapi_name = container_name(node_name, "ethapi");
+
+        let fendermint = DockerContainer::get(&dh, fendermint_name.clone()).await?;
+        let cometbft = DockerContainer::get(&dh, cometbft_name.clone()).await?;
+        let ethapi = DockerContainer::get(&dh, ethapi_name.clone()).await?;
+
+        // Directory for the node's data volumes
+        let node_dir = root.as_ref().join(node_name);
+        std::fs::create_dir_all(&node_dir).context("failed to create node dir")?;
+
+        // Get the current user ID to use with docker containers.
+        let user = node_dir.metadata()?.uid();
+
+        // Create a directory for keys
+        let keys_dir = node_dir.join("keys");
+        if !keys_dir.exists() {
+            std::fs::create_dir(&keys_dir)?;
+        }
+
+        // Create a directory for cometbft
+        let cometbft_dir = node_dir.join("cometbft");
+        if !cometbft_dir.exists() {
+            std::fs::create_dir(&cometbft_dir)?;
+
+            // We'll need to run some cometbft and fendermint commands.
+            // NOTE: Currently the Fendermint CLI commands live in the
+            // `app` crate in a way that they can't be imported. We
+            // could move them to the `lib.rs` from `main.rs` and
+            // then we wouldn't need docker for some of these steps.
+            // However, at least this way they are tested.
+
+            let cometbft_runner = DockerRunner::new(
+                &dh,
+                node_name,
+                COMETBFT_IMAGE,
+                user,
+                vec![(cometbft_dir.clone(), "/cometbft")],
+            );
+
+            let fendermint_runner = DockerRunner::new(
+                &dh,
+                node_name,
+                FENDERMINT_IMAGE,
+                user,
+                vec![
+                    (keys_dir.clone(), "/fendermint/keys"),
+                    (cometbft_dir.clone(), "/cometbft"),
+                    (node_config.genesis.path.clone(), "/fendermint/genesis.json"),
+                ],
+            );
+
+            // Init cometbft to establish the network key.
+            cometbft_runner
+                .run_cmd("init")
+                .await
+                .context("cannot init cometbft")?;
```

Comment:
in case any of these fail, should we delete the `cometbft_dir` so it will not skip this step on resume?

Reply:
Yeah, that's a good point! I'll do it in the next PR, had to fix many things with this.

Reply:
Fixed in https://github.com/consensus-shipyard/ipc/pull/744/commits/e3858ff082e0017ca43be6c8322e32b2b4fd26c4